### PR TITLE
Add account recovery setting section on the security settings screen.

### PR DESCRIPTION
**Pull request by thivi**
Create a new Settings section called Account Recovery. Move the Security Questions section into the Account Recovery section. Add Email Recovery option to Account Recovery.
image

### CODE REVIEW

File: apps/user-portal/src/components/account-recovery-email.tsx [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+    };
+
+    const showEditView = () => {
+        if (!isEdit) {
+            return (<Grid padded>
+                    <Grid.Row columns={2}>
+                        <Grid.Column width={11} className="first-column">
+                            <List.Content floated="left">
+                                <ThemeIcon
+                                    icon={MFAIcons.sms}
```
Comment (NipuniBhagya) [inline apps/user-portal/src/components/account-recovery-email.tsx]: Let's have a separate icon for email recovery

File: apps/user-portal/src/components/account-recovery-code.tsx [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    };
+
+    const showEditView = () => {
+        if (!isEdit) {
+            return (<Grid padded>
+                    <Grid.Row columns={2}>
+                        <Grid.Column width={11} className="first-column">
+                            <List.Content floated="left">
+                                <ThemeIcon
+                                    icon={MFAIcons.sms}
```
Comment (NipuniBhagya) [inline apps/user-portal/src/components/account-recovery-code.tsx]: Let's add a separate icon for recovery code

File: apps/user-portal/src/components/account-recovery-code.tsx [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                            "views:userProfile.notification.getProfileInfo.error.message"
+                        ),
+                        otherProps: {
+                            negative: true
+                        },
+                        visible: true
+                    });
+                }
+            });
+    };
```
Comment (NipuniBhagya) [inline apps/user-portal/src/components/account-recovery-code.tsx]: Why are we updating the phone number here?

File: apps/user-portal/src/components/index.ts [hunk 4]
```diff
@@ -28,11 +28,14 @@ export * from "./profile-export";
...
 export * from "./user-sessions";
 export * from "./icon";
 export * from "./consent-management";
 export * from "./side-panel";
 export * from "./pending-approvals";
 export * from "./user-image";
+export * from "./account-recovery";
+export * from "./account-recovery-email";
+export * from "./account-recovery-code";
+export * from "./account-recovery-security-qs";
```
Comment (NipuniBhagya) [inline apps/user-portal/src/components/index.ts]: Please add a new line here

File: apps/user-portal/src/components/account-recovery-security-qs.tsx [hunk 5]
```diff
@@ -89,11 +91,11 @@ class SecurityQuestionsComponentInner extends React.Component<ComponentProps, an
      * question set ids of the questions fetched from the api.
      */
     public initModel = () => {
-        const {challenges, challengeQuestions} = this.state;
-        const challengesCopy = [{...challengeQuestions}];
+        const { challenges, challengeQuestions } = this.state;
```
Comment (NipuniBhagya) [inline apps/user-portal/src/components/account-recovery-security-qs.tsx]: We only need to add a space between the curly brackets and the characters in the imports. Shall we remove this space in the constants declaration?

File: apps/user-portal/src/components/account-recovery-security-qs.tsx [hunk 6]
```diff
@@ -248,54 +250,72 @@ class SecurityQuestionsComponentInner extends React.Component<ComponentProps, an
...
-                                );
-                            })
-                        }
-                    </List>
+                    <Grid padded>
+                        <Grid.Row columns={2}>
+                            <Grid.Column width={11} className="first-column">
+                                <List.Content floated="left">
+                                    <ThemeIcon
+                                        icon={MFAIcons.sms}
```
Comment (NipuniBhagya) [inline apps/user-portal/src/components/account-recovery-security-qs.tsx]: shall we have a separate icon for security questions too?

File: apps/user-portal/src/components/account-recovery-code.tsx [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+import { useTranslation } from "react-i18next";
+import { Button, Divider, Form, Grid, Icon, List } from "semantic-ui-react";
+import { getProfileInfo, updateProfileInfo } from "../actions";
+import { MFAIcons } from "../configs";
+import { NotificationActionPayload } from "../models/notifications";
+import { EditSection } from "./edit-section";
+import { ThemeIcon } from "./icon";
+
+/**
+ * Proptypes for the SMS OTP component.
```
Comment (jeradrutnam) [inline apps/user-portal/src/components/account-recovery-code.tsx]: Has to be "Recover by code"

File: apps/user-portal/src/components/account-recovery-code.tsx [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        };
+
+        data.Operations[0].value = {
+            phoneNumbers: [
+                {
+                    type: "mobile",
+                    value: editedMobile
+                }
+            ]
+        };
```
Comment (jeradrutnam) [inline apps/user-portal/src/components/account-recovery-code.tsx]: This shouldn't be here. It should have the recover by code logic

File: apps/user-portal/src/components/account-recovery-code.tsx [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+
+/**
+ * Proptypes for the SMS OTP component.
+ */
+interface CodeRecoveryProps {
+    onNotificationFired: (notification: NotificationActionPayload) => void;
+}
+
+/**
+ * SMS OTP section.
```
Comment (NipuniBhagya) [inline apps/user-portal/src/components/account-recovery-code.tsx]: This needs to be changed to "Recover by code section" or something similar.

File: apps/user-portal/src/components/account-recovery-code.tsx (hunk 3 above)
Comment (NipuniBhagya) [inline apps/user-portal/src/components/account-recovery-code.tsx]: I think we don't need this logic here.

File: apps/user-portal/src/components/account-recovery-code.tsx [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+    };
+
+    const setMobileNo = (resp) => {
+        let mobileNumber = "";
+        resp.phoneNumbers.map((mobileNo) => {
+            mobileNumber = mobileNo.value;
+        });
+        setMobile(mobileNumber);
+        setEditedMobile(mobileNumber);
+    };
```
Comment (NipuniBhagya) [inline apps/user-portal/src/components/account-recovery-code.tsx]: shall we remove this part as well?

File: apps/user-portal/src/components/account-recovery-email.tsx [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+        setEditedEmail(event.target.value);
+        event.preventDefault();
+    };
+
+    /**
+     * This function gets the email address from the response passed as the argument 
+     * and assigns it to email and editedEmail. 
+     * @param resp 
+     */
+    const setEmailAddr = (resp) => {
```
Comment (jeradrutnam) [inline apps/user-portal/src/components/account-recovery-email.tsx]: Use `setEmailAddress` instead

File: apps/user-portal/src/components/account-recovery-email.tsx [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+        let emailAddress = "";
+        resp.emails
```
Comment (jeradrutnam) [inline apps/user-portal/src/components/account-recovery-email.tsx]: Use `response` instead `resp`

File: apps/user-portal/src/components/account-recovery-email.tsx [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+    useEffect(() => {
+        getProfileInfo()
+            .then((response) => {
+                setEmailAddr(response);
+            });
+
+    }, []);
+
+    return (
+        <div>
```
Comment (jeradrutnam) [inline apps/user-portal/src/components/account-recovery-email.tsx]: Is this additional `div` tag needed ? if not just remove it

File: apps/user-portal/src/components/account-recovery-email.tsx [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+            </EditSection>
+        );
+    };
+
+    useEffect(() => {
+        getProfileInfo()
+            .then((response) => {
+                setEmailAddr(response);
+            });
+
```
Comment (jeradrutnam) [inline apps/user-portal/src/components/account-recovery-email.tsx]: Handle the `return` here. Else there will be continuous call to the method. As `useEffect` hook uses on both `componentDidMount` & `componentDidUpdate`

File: apps/user-portal/src/components/account-recovery-email.tsx [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+     * This function returns the EditSection component and the recovery option 
+     * elements based on if the edit icon has been clicked
+     */
+    const showEditView = () => {
+        if (!isEdit) {
+            return (<Grid padded>
+                <Grid.Row columns={2}>
+                    <Grid.Column width={11} className="first-column">
+                        <List.Content floated="left">
+                            <ThemeIcon
```
Comment (NipuniBhagya) [inline apps/user-portal/src/components/account-recovery-email.tsx]: Can we align this Icon with the list title?

File: apps/user-portal/src/components/account-recovery-email.tsx (hunk 1 above)
Comment (thivi) [reply apps/user-portal/src/components/account-recovery-email.tsx]: This has been rectified. 
image

File: apps/user-portal/src/components/account-recovery-code.tsx (hunk 2 above)
Comment (thivi) [reply apps/user-portal/src/components/account-recovery-code.tsx]: The recovery code section has been removed for the time being.

File: apps/user-portal/src/components/index.ts (hunk 4 above)
Comment (thivi) [reply apps/user-portal/src/components/index.ts]: This has been resolved.

File: apps/user-portal/src/components/account-recovery-security-qs.tsx (hunk 6 above)
Comment (thivi) [reply apps/user-portal/src/components/account-recovery-security-qs.tsx]: A new icon has been added for security questions. 
image

File: apps/user-portal/src/components/account-recovery-code.tsx (hunk 3 above)
Comment (thivi) [reply apps/user-portal/src/components/account-recovery-code.tsx]: This section has been removed.

Comment (thivi) [reply apps/user-portal/src/components/account-recovery-code.tsx]: This section has been removed.

File: apps/user-portal/src/components/account-recovery-email.tsx (hunk 13 above)
Comment (thivi) [reply apps/user-portal/src/components/account-recovery-email.tsx]: The tag is not needed and has been removed.

File: apps/user-portal/src/components/account-recovery-email.tsx (hunk 15 above)
Comment (thivi) [reply apps/user-portal/src/components/account-recovery-email.tsx]: The icon is now aligned correctly. 
image

File: apps/user-portal/src/components/account-recovery-email.tsx (hunk 12 above)
Comment (thivi) [reply apps/user-portal/src/components/account-recovery-email.tsx]: The change has been made.

File: apps/user-portal/src/components/account-recovery-code.tsx (hunk 10 above)
Comment (thivi) [reply apps/user-portal/src/components/account-recovery-code.tsx]: The whole section has been removed.

File: modules/theme/src/themes/default/assets/images/icons/email-icon.svg [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+-->
+<svg xmlns:svg="http://www.w3.org/2000/svg" 
+   xmlns="http://www.w3.org/2000/svg" id="email-icon" height="14" width="14" version="1.1" xml:space="preserve" viewBox="0 0 420 420" y="0px" x="0px" version="1.1" class="icon">
+
```
Comment (jeradrutnam) [inline modules/theme/src/themes/default/assets/images/icons/email-icon.svg]: Remove additional new line

File: modules/theme/src/themes/default/assets/images/icons/email-icon.svg [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+   <path class="path primary" d="M 209.46992,91.817395 208.35184,93.095736 11.863524,319.59479 H 407.07633 Z m 0,4.532303 191.07445,220.281642 
+   H 18.39548 Z" />
+   <polygon class="path primary" transform="matrix(30.129386,0,0,29.750516,-1.4357757,-47.521938)" points="13.45,2.22 7,9.76 0.55,2.22 " />
+
+   <path class="path secondary" d="M 38.324416,0.05859375 C 17.194139,0.05859375 0.07421875,17.362777 0.07421875,37.885909 V 295.18139 c 0,20.86455 
+   17.46570225,37.82731 38.25019725,37.82731 H 380.63425 c 21.02362,0 38.25019,-17.06808 38.25019,-37.82731 V 37.885909 c 0,-20.864566 -17.4657,
+   -37.82731525 -38.25019,-37.82731525 z m 0,29.75051725 H 380.63425 c 4.5242,0 8.1208,3.355924 8.1208,8.076798 V 295.18139 c 0,4.23116 -3.83574,8.0768 
+   -8.1208,8.0768 H 38.324416 c -4.524193,0 -8.12081,-3.35595 -8.12081,-8.0768 V 37.885909 c 0,-4.467296 3.339812,-8.076798 8.12081,-8.076798 z" />
+
```
Comment (jeradrutnam) [inline modules/theme/src/themes/default/assets/images/icons/email-icon.svg]: Remove additional new line

File: modules/theme/src/themes/default/assets/images/icons/security-questions-icon.svg [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+-->
+<svg xmlns:svg="http://www.w3.org/2000/svg" 
+	xmlns="http://www.w3.org/2000/svg" id="security-questions-icon" height="14" width="14" xml:space="preserve" viewBox="0 0 420 420" y="0px" x="0px" id="Layer_1" version="1.1" class="icon">
+
```
Comment (jeradrutnam) [inline modules/theme/src/themes/default/assets/images/icons/security-questions-icon.svg]: Remove additional new line

File: apps/user-portal/src/components/account-recovery-email.tsx [hunk 19]
```diff
@@ -0,0 +1,270 @@
...
+                                    rounded
+                                    relaxed
+                                />
+                            </List.Content>
+                            <List.Content>
+                                <List.Header>{t("views:accountRecovery.emailRecovery.title")}</List.Header>
+                                <List.Description>
+                                    {
+                                        email || email != ""
+                                            ? t("views:accountRecovery.emailRecovery.descriptionUpdate", { email: maskEmail(email) })
```
Comment (jeradrutnam) [inline apps/user-portal/src/components/account-recovery-email.tsx]: Fix character length

File: modules/theme/src/themes/default/assets/images/icons/email-icon.svg (hunk 16 above)
Comment (thivi) [reply modules/theme/src/themes/default/assets/images/icons/email-icon.svg]: All empty lines have been removed

File: modules/theme/src/themes/default/assets/images/icons/email-icon.svg (hunk 17 above)
Comment (thivi) [reply modules/theme/src/themes/default/assets/images/icons/email-icon.svg]: All empty lines have been removed.

File: apps/user-portal/src/components/account-recovery-email.tsx (hunk 19 above)
Comment (thivi) [reply apps/user-portal/src/components/account-recovery-email.tsx]: This issue has been fixed.

**Pull request merged**